### PR TITLE
Added DagshubCallback

### PR DESCRIPTION
**Pull request by jinensetpal**
# What does this PR do?

Adds a Trainer Integration with [DagsHub](https://dagshub.com/). It extends the MLFlow integration that already currently exists to also track and push artifacts using DVC, into DagsHub repositories. The idea is to allow users to integrate an MLOps stack to their current setups with minimal effort.

Here's a colab notebook with a sample integration: https://colab.research.google.com/drive/1KEeQYp3jD8kmkMMhGKR1R1C6cEjnbJLZ

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/integrations.py [hunk 1]
```diff
@@ -1045,6 +1051,52 @@ def __del__(self):
...
 
+class DagsHubCallback(MLflowCallback):
+    """
+    A [`TrainerCallback`] that logs to [DagsHub](https://dagshub.com/).
+    """
+
+    def __init__(self):
+        super().__init__()
+        if not is_dagshub_available():
+            raise RuntimeError("DagsHubCallback requires dagshub to be installed. Run `pip install dagshub`.")
```
Comment (sgugger) [inline src/transformers/integrations.py:1062]:
```suggestion
            raise ImportError("DagsHubCallback requires dagshub to be installed. Run `pip install dagshub`.")
```

File: src/transformers/integrations.py [hunk 2]
```diff
@@ -1045,6 +1051,52 @@ def __del__(self):
...
+        from glob import glob
```
Comment (sgugger) [inline src/transformers/integrations.py:1063]: This is standard lib, this import can go on top of the file.

File: src/transformers/integrations.py [hunk 3]
```diff
@@ -1045,6 +1051,52 @@ def __del__(self):
...
+
+        Environment:
+            HF_DAGSHUB_LOG_ARTIFACTS (`str`, *optional*):
+                Whether to save the data and model artifacts for the experiment. Default to `False`.
+        """
+
+        self.model = args[2]
+        self.log_artifacts = os.getenv("HF_DAGSHUB_LOG_ARTIFACTS", "FALSE").upper() in ENV_VARS_TRUE_VALUES
+        self.name = os.getenv("HF_DAGSHUB_MODEL_NAME") or "main"
+        self.remote = os.getenv("MLFLOW_TRACKING_URI") or input("Please input a remote url.")
```
Comment (sgugger) [inline src/transformers/integrations.py:1082]: We can't use input in a training script that may be launched in several processes. You should error if this env variable is not set.

**Review by sgugger — COMMENTED** [review]
Thanks for adding this! Maybe the logic in `get_available_reporting_integrations` will need to be updated as there is going to be two MlflowCallback in the list if `dagshub` is installed. Is that going to be a problem?

**Comment by jinensetpal** [comment]
Thank you so much for the lightning review, @sgugger!

I've made the relevant changes. I also added a check within `get_available_reporting_integrations` to ensure available integrations are reported accurately. 

Since `DagsHubCallback` subclasses `MLFlowCallback`, it's still separate. If both integrations are requested, it does not execute twice. This is what the `get_available_reporting_integrations` looks like when moving packages around:

2023-02-01-095908_1920x606_scrot

Hope this was helpful! :)

File: src/transformers/integrations.py [hunk 4]
```diff
@@ -1380,7 +1435,6 @@ def setup(self, args, state, model, tokenizer, **kwargs):
             if self._clearml_task is None:
                 self._clearml_task = self._clearml.Task.init(
                     project_name=os.getenv("CLEARML_PROJECT", "HuggingFace Transformers"),
-                    task_name=os.getenv("CLEARML_TASK", "Trainer"),
```
Comment (sgugger) [inline src/transformers/integrations.py:1383]: Not sure why there is this change added here.

File: src/transformers/integrations.py [hunk 5]
```diff
@@ -1045,6 +1051,55 @@ def __del__(self):
...
+        self.name = os.getenv("HF_DAGSHUB_MODEL_NAME") or "main"
+        self.remote = os.getenv("MLFLOW_TRACKING_URI")
+        self.repo = self.Repo(
+            owner=self.remote.split(os.sep)[-2],
+            name=self.remote.split(os.sep)[-1].split(".")[0],
+            branch=os.getenv("BRANCH") or "main",
+        )
+        self.path = Path("artifacts")
+
+        if not self.remote:
```
Comment (sgugger) [inline src/transformers/integrations.py:1087]:
```suggestion
        if self.remote is None:
```
We don't rely on Python bool conversion magic in Transformers and prefer explicit tests.

**Review by sgugger — COMMENTED** [review]
Thanks for confirming! Two more comments and this should be good to merge!

File: src/transformers/integrations.py (hunk 4 above)
Comment (jinensetpal) [reply src/transformers/integrations.py:1383]: That's an accident - it shouldn't have been removed! Reverting right away.

**Review by sgugger — APPROVED** [review]

**Comment by sgugger** [comment]
Thanks again for your contribution!

**Comment by jinensetpal** [comment]
Thank you for your rapid responses!! Have an awesome rest of day 🙂

**Pull request merged**